### PR TITLE
Custom expression editor: use icons to distinguish the suggestions

### PR DESCRIPTION
**Pull request by ariya**
Still WIP.

How to try:
1. Ask a question, Custom question
2. Sample Dataset, Product table
3. Filter, Custom expression, and type `c`

**Before this PR**

The suggestions are separated by their types, e.g. `FIELDS`, `FUNCTIONS`, and `SEGMENTS` in the following screenshot.

image
**After this PR**

There is no explicit separator. Each suggestion has a corresponding icon indicating its type.

image

### CODE REVIEW

**Review by flamber — APPROVED** [review]
Wow - this is insanely useful compared to currently, where I have to guess some column types:
image

Wonder if #18537 is even needed anymore, now that it's possible to just click the field an see it's a Segment. Alternatively, using icons here (might be way too ugly/distracting)?

File: frontend/src/metabase/icon_paths.ts [hunk 1]
```diff
@@ -201,6 +201,8 @@ export const ICON_PATHS: Record<string, any> = {
     "M0,0 L7.51851852,0 L7.51851852,3.2 L0,3.2 L0,0 Z M10.7407407,0 L18.2592593,0 L18.2592593,3.2 L10.7407407,3.2 L10.7407407,0 Z M21.4814815,0 L29,0 L29,3.2 L21.4814815,3.2 L21.4814815,0 Z M0,5.33333333 L7.51851852,5.33333333 L7.51851852,29.8666667 L3.85540334,32 L0,29.8666667 L0,5.33333333 Z M10.7407407,5.33333333 L18.2592593,5.33333333 L18.2592593,29.8666667 L14.5690136,32 L10.7407407,29.8666667 L10.7407407,5.33333333 Z M21.4814815,5.33333333 L29,5.33333333 L29,29.8666667 L25.2114718,32 L21.4814815,29.8666667 L21.4814815,5.33333333 Z",
   filter:
     "M3.556 4h24.888a3.556 3.556 0 1 1 0 7.111H3.556a3.556 3.556 0 1 1 0-7.111zm4 11.556h16.888a2.222 2.222 0 1 1 0 4.444H7.556a2.222 2.222 0 0 1 0-4.444zM12 24.444h8a2.222 2.222 0 0 1 0 4.445h-8a2.222 2.222 0 0 1 0-4.445z",
+  function:
```
Comment (ariya) [inline frontend/src/metabase/icon_paths.ts:204]: Many thanks to @mazameli for this icon!

**Comment by ariya** [comment]
> Wonder if #18537 is even needed anymore, now that it's possible to just click the field an see it's a Segment. Alternatively, using icons here (might be way too ugly/distracting)?

Icons might be too distracting indeed, but IMO colors will be still good to have (and yes, the colors in the editor need to be consistent with the suggestions, e.g. purple for segments).

**Review by mazameli — CHANGES_REQUESTED** [review]
I only have 1 real change request: like the mocks show, I'd like ot have the rest of a suggestion's text be set to `color: text-medium` to further contrast it from the portion of the suggestion's text that matches what the user has typed so far:

image

Additionally, my thought had been to put functions above columns, whereas I see in the branch that this is the opposite. I don't have strong conviction on this point, however, and would love to hear @flamber's, @ariya's, and perhaps @ukritw's thoughts on the order of suggestions. (The only thing I'm pretty sure of is that "Metrics" and "Segments" in their current form aren't super useful, and I'm glad that we've put them at the bottom.)

image

Also, just to say it: I noticed that the function icon is colored red rather than blue like in the mockups, but I kind of prefer the red now that I'm seeing it — but was just curious if that was an intentional choice.

**Comment by flamber** [comment]
It might be a relic with the order, but something tells me that you're inputting more Fields than Functions, so it makes sense that Fields are first, though from a logical point, Functions should probably be first.

I think I've generally been annoyed no matter what, since when I have an empty expression, then I usually need a function, but inside a function I usually need a field. But it would be even worse if we tried to do some auto-guess, so consistency wins for me.

**Comment by ariya** [comment]
> I only have 1 real change request: like the mocks show, I'd like ot have the rest of a suggestion's text be set to `color: text-medium` to further contrast it from the portion of the suggestion's text that matches what the user has typed so far:

That makes sense, looks like I forgot to tweak the text color.


> Additionally, my thought had been to put functions above columns, whereas I see in the branch that this is the opposite. I don't have strong conviction on this point, however, and would love to hear @flamber's, @ariya's, and perhaps @ukritw's thoughts on the order of suggestions. (The only thing I'm pretty sure of is that "Metrics" and "Segments" in their current form aren't super useful, and I'm glad that we've put them at the bottom.)

I don't have any strong preference here, I defer it to others to offer some more meaningful suggestions.

> Also, just to say it: I noticed that the function icon is colored red rather than blue like in the mockups, but I kind of prefer the red now that I'm seeing it — but was just curious if that was an intentional choice.

It was an accident, I think I picked the wrong accent.
Blue does make sense since that is the color used for the syntax highlighting (once that function name appears in the editor), so we should strive for consistency here.

**Comment by ariya** [comment]
@mazameli Please check again, I've made some color tweaks.

**Comment by ukritw** [comment]
I have no strong preference in the ordering either. Like @flamber said, consistency is key.

**Comment by mazameli** [comment]
This is looking almost there to me. The icons in the field suggestions though are looking a bit small, particularly the Category `Aa` and the Number `#` ones. I specced them to be 14px, whereas I think they're 10px in the branch.

image

Here's the design mockup for comparison:

image

**Comment by ariya** [comment]
> This is looking almost there to me. The icons in the field suggestions though are looking a bit small, particularly the Category `Aa` and the Number `#` ones. I specced them to be 14px, whereas I think they're 10px in the branch.

Good catch @mazameli!
I've changed it to 14px, but now we have some (vertical) alignment issue. Can I get some CSS help on this?

image

**Review by mazameli — APPROVED** [review]

**Comment by mazameli** [comment]
@ariya I pushed a little CSS fix:

image

**Comment by ariya** [comment]
> I have no strong preference in the ordering either. Like @flamber said, consistency is key.

A minor consistency argument in favor of listing the functions first is making the the list a bit more _**predictable**_, e.g. there are only 3 functions that start with an `s` (sqrt, startsWith, substring), whereas the fields and others vary significantly depending on the table etc.

**Comment by akiselev** [comment]
> > I have no strong preference in the ordering either. Like @flamber said, consistency is key.
> 
> A minor consistency argument in favor of listing the functions first is making the the list a bit more _**predictable**_, e.g. there are only 3 functions that start with an `s` (sqrt, startsWith, substring), whereas the fields and others vary significantly depending on the table etc.

How is autocomplete used the most and which scenario are we optimizing for? Is it for discoverability/mental unblocking the average user as they explore data or is it used by power users to autocomplete long names using muscle memory?

**Comment by ariya** [comment]
> How is autocomplete used the most and which scenario are we optimizing for? Is it for discoverability/mental unblocking the average user as they explore data or is it used by power users to autocomplete long names using muscle memory?

@flamber and @mazameli  will know better, though I suspect that there won't be a lot of non-power users dropping to the custom expression editor since they can already perform a number of complex filtering and aggregation via the simpler UI.

**Pull request merged**